packages/aragon-cli/src/helpers/solidity-extractor.js

```diff
@@ -84,21 +85,71 @@ const getRoles = declaration => {
   )
 }
 
-// Takes the path to a solidity file and extracts public function signatures,
-// its auth role if any and its notice statement
-module.exports = async sourceCodePath => {
-  const sourceCode = await readFile(sourceCodePath, 'utf8')
-
-  // everything between every 'function' and '{' and its @notice
-  const funcDecs = sourceCode.match(/(@notice|^\s*function)(?:[^]*?){/gm)
+const extractFunctions = async (sourceCode) => {
+  // Everything between every 'function' and '{' and its @notice.
+  const functionDeclarations = sourceCode.match(/(@notice|^\s*function)(?:[^]*?){/gm)
 
-  if (!funcDecs) return []
+  if (!functionDeclarations) {
+    return []
+  }
 
-  return funcDecs
-    .filter(dec => modifiesStateAndIsPublic(dec))
-    .map(dec => ({
-      sig: getSignature(dec),
-      roles: getRoles(dec),
-      notice: getNotice(dec),
+  const stateModifyingFunctionDeclarations = functionDeclarations
+    .filter(functionDeclaration => modifiesStateAndIsPublic(functionDeclaration))
+    .map(functionDeclaration => ({
+      sig: getSignature(functionDeclaration),
+      roles: getRoles(functionDeclaration),
+      notice: getNotice(functionDeclaration),
     }))
+
+  return stateModifyingFunctionDeclarations
 }
+
+const extractRoles = async (functionDescriptors) => {
+  // Extract all role ids from the function descriptors.
+  let roleSet = new Set()
+  functionDescriptors.forEach(({ roles }) => roles.forEach(role => roleSet.add(role)))
+  const roleIds = [...roleSet]
+
+  // Parse role ids into objects.
+  // TODO: Name and parameters are currently not being extracted, and it's probably better to get it from an AST instead of the Solidity code. For now, the properties are merely place holders.
+  return roleIds.map(id => ({
+    id,
+    bytes: '0x' + keccak256(id),
```

Comment:
```suggestion
    bytes: keccak256(id),
```

Reply:
Maybe we can do this in a safe way such as
```js
const hash = hashFunction(id)
return "0x" + (hash.split("0x")[1] || hash)
```
or
```js
const hash = hashFunction(id)
return hash.startsWith("0x") ? hash : "0x" + hash
```
If it's unclear what return format it does

Reply:
@dapplion Yes I think it could be a good idea to abstract this away in case we switch again to another `keccak256` function that doesn't include `0x`. I will create a separate issue for this since the function is used at a lot of different places other than this PR.

Reply:
See #914